doc/rfc008-managing-apalache.md

```diff
@@ -0,0 +1,343 @@
+
+Pros:
+
+- Independent of our current release/packaging practice.
+- We can simply state version dependencies in our docs or release notes;
+  resolving them lies with the packager.
+- Transparent to the user.
+- Reduced complexity for Quint / the VSCode extension: no need to install /
+  manage Apalache from there.
+
+Cons:
+
+- Only leaves one option for resolving §3.2: Since this requires
+  platform-dependent adaptions to register as a system-level service, this
+  solution largely depends on installing Quint and Apalache through a
+  system-level package manager (§3.2.1), and inherits its pros/cons.
+- Shai permanently takes up resources (freshly launched: 125MB memory), while
+  it's likely only used intermittently.
+- Requires platform-dependent error message / intervention if Shai is not
+  running.
+- Additional logic in Quint for checking version compatilibity with Apalache.
+
+#### 3.1.2 Manually by the user (in a separate terminal, ...)
+
+This is our current setup for development practice: Shai is manually launched by
+the user, usually in a separate terminal / screen session / or similar.
+
+Pros:
+
+- Independent of §3.2, as long as the `apalache-mc` command is available.
+- Transparent to the user, running `apalache-mc server` is platform-independent.
+- Reduced complexity for Quint / the VSCode extension: no need to install /
+  manage Apalache from there.
+
+Cons:
+
+- Increased user overhead, for a task that could be automated.
+- Additional logic in Quint for checking version compatilibity with Apalache.
+
+##### Launching of a system service by the user
+
+
+As a special case, the user may manually launch a system service (§3.1.1
+above), e.g., via `systemctl` on Linux or `launchctl` on macOS. The pros/cons
+in this case fall at the intersection of both options.
+
+#### 3.1.3 Managed by Quint
+
+This option launches Shai from the Quint binary (or the VSCode extension).
+
+This is similar to how bazel [runs its build server](https://bazel.build/run/client-server),
+and to how the Metals VSCode plugin [manages the metals server](https://github.com/scalameta/metals-vscode/blob/edc446cc58dc809f3ed7cfde15b1e7fd7b5be115/packages/metals-vscode/src/extension.ts#L175C10-L175C30).
+
+##### On-demand vs long-lived
+
+As a sub-option, we would have to decide if Shai is launched on-demand, for each
+verification invocation, or long-lived for as long as the VSCode extension is
+active. bazel and vscode-metals employ a hybrid approach, where a long-lived
+server is launched on-demand and torn down after an inactivity timeout. This
+decision should take into consideration the memory-overhead of a long-lived
+server vs. JVM startup time for an on-demand approach; at the time of writing,
+a freshly launched `apalache-mc server` takes ~125MB of memory, while JVM
+startup time is negligable (< 2sec).
```

Comment:
Relative to usual Apalache runtime, I think it is?

Reply:
I guess if it was really negligible, then there would be no reason for us to consider managing the server in any scenario. Perhaps
```suggestion
server vs. JVM startup time for an on-demand approach; at the time of writing,
a freshly launched `apalache-mc server` takes ~125MB of memory, while JVM
startup time is negligible for most runs (< 2sec).
```

One example scenario where this is not negligible may be running the apalache simulator with a few runs to iterate over something (once we have a command for that)

Reply:
> then there would be no reason for us to consider managing the server in any scenario

ergonomics, I think

> running the apalache simulator with a few runs to iterate over something

I think what you describe here is still a single invocation of the simulate command in Apalache? Otherwise, we'd probably want the more fine-grained interface described in [Apalache RFC-010](https://apalache.informal.systems/docs/adr/010rfc-transition-explorer.html)?

Reply:
If you are running 10

Reply:
It looks to me that there are two main user stories here:

 1. A first time user tries to run `quint verify`. They want to try the command and definitely do not want to manage the server (they probably even do not know what it means).
 2. An experienced user runs `quint verify` periodically. They would prefer to have (one instance of) the server in the background.

At this point of development, I would prioritise the first-time user over the experienced user. However, we should give the experienced user an option to optimize their workflow by running the server in the background. As far as I understand, there is such an option.